### PR TITLE
Add a custom switch to clean up the leaked Dart VM

### PR DESCRIPTION
**Pull request by eggfly**
It's both Android and iOS part of:
* https://github.com/flutter/flutter/issues/95903

Considering the `leak_vm` setting is  an process global option, I put it into AndroidManifest.xml's app meta-data, like the `OldGenHeapSize` option. 

Demo app link:  https://github.com/eggfly/flutter_leak_vm_test

Code:

https://github.com/eggfly/flutter_leak_vm_test/blob/612b2271c90de68f9bbaf03696fcdb18bcd73745/android/app/src/main/AndroidManifest.xml#L51

```xml
 <!-- Force clean up the leak VM -->
        <meta-data
            android:name="io.flutter.embedding.android.LeakVM"
            android:value="false" />
```

Keep switching between native MainActivity and flutter activity, and observe the memory usage.

LeakVM is true (or not set): back to native Activity then native memory = 34.9MB
image

LeakVM is false: back to native Activity then native memory = 32.2MB
image

Some debug logs:

image

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: runtime/dart_vm_lifecycle.cc [hunk 1]
```diff
@@ -41,7 +41,7 @@ DartVMRef::~DartVMRef() {
   vm_.reset();
 }
 
-DartVMRef DartVMRef::Create(Settings settings,
+DartVMRef DartVMRef::Create(const Settings& settings,
```
Comment (chinmaygarde) [inline runtime/dart_vm_lifecycle.cc:44]: This seems like a drive by change unrelated to the patch.

File: shell/common/switches.h [hunk 2]
```diff
@@ -215,6 +215,12 @@ DEF_SWITCH(EnableSkParagraph,
            "enable-skparagraph",
            "Selects the SkParagraph implementation of the text layout engine.")
 
+DEF_SWITCH(RetainVM,
+           "retain-vm",
```
Comment (chinmaygarde) [inline shell/common/switches.h:219]: Perhaps a pet peeve of mine, but the "retain" may indicate that that there is a way to release the VM later. There is no such feature. Also, calling this retain-vm in one place and the same thing leak-vm in another is inconsistent and potentially confusing. I would just call it leak-vm like the shell switch.

**Review by chinmaygarde — COMMENTED** [review]
cc @blasten for whether this we should do this. I think, if anything, we should switch the default on the flag to never leak the VM.

Ideally, this flag should not exist and the last instance of the shell should always collect the VM. But this was added because of benchmark regressions on application resumption. IMO, we should migrate to non-leaky behavior instead of exposing another flag.

File: runtime/dart_vm_lifecycle.cc (hunk 1 above)
Comment (eggfly) [reply runtime/dart_vm_lifecycle.cc]: Removed them. Done.

File: shell/common/switches.h (hunk 2 above)
Comment (eggfly) [reply shell/common/switches.h]: Reasonable & agree with you. Keep them all called "leak" as same ,changed these var to leak_vm/ LeakVM. Done

File: shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java [hunk 3]
```diff
@@ -318,6 +327,14 @@ public void ensureInitializationComplete(
     }
   }
 
+  private static boolean isLeakVM(Bundle metaData) {
```
Comment (blasten) [inline shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java:330]:
```suggestion
  private static boolean isLeakVM(@Nullable Bundle metaData) {
```

**Comment by blasten** [comment]
@goderbauer @dnfield 
do you know if `customer:money` needs this flag to be on by default?

Some context about this flag: https://github.com/flutter/engine/blob/ea734ba7f8de9c6e2bb6b30efa1f53b5aec6d3c6/common/settings.h#L198-L209

**Comment by blasten** [comment]
This change seems reasonable. I also think that the default behavior should be to never leak the VM. Do you know how this interacts with cached engines, e.g. https://api.flutter.dev/javadoc/io/flutter/embedding/engine/FlutterEngineCache.html ?

**Comment by dnfield** [comment]
This flag was specifically set up because on Android and iOS applications, going to and from background was causing the VM to restart. We don't want that to happen because it is slow to restart the VM.

**Comment by eggfly** [comment]
> This flag was specifically set up because on Android and iOS applications, going to and from background was causing the VM to restart. We don't want that to happen because it is slow to restart the VM.

I agree with you. And also add the `setLeakVM` method for iOS, comment it to `YES` for default.

**Comment by blasten** [comment]
Let's land this change with the current default.

Please file an issue to change the default value, and add TODO with the link to the issue.

Another PR can change the default to `false`, so it can be tested separately.

File: shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java [hunk 4]
```diff
@@ -40,6 +40,12 @@
   private static final String ENABLE_SKPARAGRAPH_META_DATA_KEY =
       "io.flutter.embedding.android.EnableSkParagraph";
 
+  /**
+   * Set whether leave or clean up the VM after the last shell shuts down. It's true by default, set
+   * it to false to destroy VM.
```
Comment (blasten) [inline shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java:45]: can you add an example use case, or more context about why someone may or may not want to leak the VM? 
e.g. is this problematic when using shared engines and/or add-to-app with some other set up?

Comment (eggfly) [reply shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java:45]: > can you add an example use case, or more context about why someone may or may not want to leak the VM? e.g. is this problematic when using shared engines and/or add-to-app with some other set up?

Can we also add more context with the comment in `settings.h` into this java and objc docs:
in `engine/common/settings.h`
```c++
 // All shells in the process share the same VM. The last shell to shutdown 
 // should typically shut down the VM as well. However, applications depend on 
 // the behavior of "warming-up" the VM by creating a shell that does not do 
 // anything. This used to work earlier when the VM could not be shut down (and 
 // hence never was). Shutting down the VM now breaks such assumptions in 
 // existing embedders. To keep this behavior consistent and allow existing 
 // embedders the chance to migrate, this flag defaults to true. Any shell 
 // launched with this flag set to true will leak the VM in the process. There 
 // is no way to shut down the VM once such a shell has been started. All 
 // shells in the platform (via their embedding APIs) should cooperate to make 
 // sure this flag is never set if they want the VM to shutdown and free all 
 // associated resources. 
```

Comment (eggfly) [reply shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java]: DONE~ Add more comments, leave a TODO and added a sample code:
* https://github.com/flutter/samples/pull/996/

File: shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm [hunk 5]
```diff
@@ -337,4 +337,8 @@ + (NSString*)defaultBundleIdentifier {
   return @"io.flutter.flutter.app";
 }
 
+- (void)setLeakDartVM:(BOOL)enabled {
+  _settings.leak_vm = enabled;
```
Comment (zhongwuzw) [inline shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm:341]: Is it better to read `leak_vm` from the main bundle's meta-plist? like https://github.com/eggfly/engine/blob/disable_leak_vm/shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm#L160-L161
```
  // SkParagraph text layout library
  NSNumber* enableSkParagraph = [mainBundle objectForInfoDictionaryKey:@"FLTEnableSkParagraph"];
  settings.enable_skparagraph = (enableSkParagraph != nil) ? enableSkParagraph.boolValue : false;
```

Keep the same way with Android. Not add a public setter method.

Comment (eggfly) [reply shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm]: Done~

**Review by zhongwuzw — APPROVED** [review]
LGTM!

**Comment by eggfly** [comment]
Added a sample code in https://github.com/flutter/samples/pull/996.

**Pull request merged**